src/core/shape_inference/include/roi_align_shape_utils.hpp

```diff
@@ -0,0 +1,106 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+#include <openvino/op/util/roi_align_base.hpp>
+
+#include "utils.hpp"
+
+namespace ov {
+namespace op {
+namespace roi_align {
+namespace validate {
+/**
+ * @brief Validates ROIs align input data and ROIs element type.
+ *
+ * @param op Pointer to ROIs align node.
+ * @return Valid ROIs align output element type.
+ */
+inline element::Type data_and_roi_et(const Node* const op) {
+    auto out_et = element::dynamic;
+
+    const auto& input_et = op->get_input_element_type(0);
+    const auto& rois_et = op->get_input_element_type(1);
+
+    NODE_VALIDATION_CHECK(op,
+                          element::Type::merge(out_et, input_et, rois_et) && out_et.is_real(),
+                          "The data type for input and ROIs is expected to be a same floating point type. Got: ",
+                          input_et,
+                          " and: ",
+                          rois_et);
+    return out_et;
+}
+
+/**
+ * @brief Check ROIs align batch indicies input element type.
+ *
+ * @param op  Pointer to ROIs align node.
+ */
+inline void batch_indicies_et(const Node* const op) {
+    const auto& indicies_et = op->get_input_element_type(2);
+
+    NODE_VALIDATION_CHECK(op,
+                          indicies_et.is_integral_number(),
+                          "The data type for batch indices is expected to be an integer. Got: ",
+                          indicies_et);
+}
+}  // namespace validate
+
+template <class TShape, class TRShape = result_shape_t<TShape>>
+std::vector<TRShape> shape_infer(const util::ROIAlignBase* op, const std::vector<TShape>& input_shapes) {
+    NODE_VALIDATION_CHECK(op, input_shapes.size() == 3);
+
+    using TDim = typename TShape::value_type;
+
+    const auto& input_ps = input_shapes[0];
+    const auto& rois_ps = input_shapes[1];
+    const auto& batch_indices_ps = input_shapes[2];
+
+    const auto rois_ps_rank = rois_ps.rank();
+    const auto input_ps_rank = input_ps.rank();
+    const auto batch_indices_ps_rank = batch_indices_ps.rank();
+
+    auto output_shapes = std::vector<TRShape>(1);
+    auto& out_shape = output_shapes.front();
+    out_shape.reserve(4);
+
+    NODE_VALIDATION_CHECK(op, input_ps_rank.compatible(4), "Expected a 4D tensor for the input data. Got: ", input_ps);
+    NODE_VALIDATION_CHECK(op, rois_ps_rank.compatible(2), "Expected a 2D tensor for the ROIs input. Got: ", rois_ps);
+    NODE_VALIDATION_CHECK(op,
+                          batch_indices_ps_rank.compatible(1),
+                          "Expected a 1D tensor for the batch indices input. Got: ",
+                          batch_indices_ps);
+
+    if (rois_ps_rank.is_static()) {
+        const auto& rois_second_dim = rois_ps[1];
+        NODE_VALIDATION_CHECK(op,
+                              rois_second_dim.compatible(op->get_rois_input_second_dim_size()),
```

Comment:
It is a part of the definition - and it is needed in shape inference. Also it is a const getter.

Reply:
Shape inference is not enough to put it in public API. It can be part only of shape inference.

Reply:
1) This is actually need for validation, which is (incorrectly) done in shape_inference - so far you had a magic numbers there.
2) It is a part of op definition - please refer to *rst files. 
3) get_pooled_h() is also used only in tests and shape inference - is it enough to make it to public API?

Reply:
Apply solution propose by @mitruska or similar.

Reply:
Guys, what you propose is simply scattering the definition of op into multiple files. This is generally speaking bad idea. Solution proposed by @mitruska is exactly the same as it already was in roi_align - add magic number but put it into separate function in separate file called 'shape_inference', **while that magic number is not needed for shape inference at all - is needed only for validation, which is incorrectly mixed up with shape inference...**

The root problem is that pattern that shape inference was separated physically from op definition - which is IMO wrong design choice. Shape inference is integral part of any op and it should be kept together. Mixing it with validation is a even worse practice IMO, but I am not going to solve all of those issue here, as it is not related to the roi_align_rotated itself. 

However I was considering another solution - making get_rois_input_second_dim_size() private and moving validation to validate_and_infer_types() in ROIAlignBase, BUT it is a refactoring of current functionality and as such - should be done in separate pr, because it is not directly related to roi align rotated.

To sum up, what I proposed here is a kind of compromise between keeping everything in one place(as it should be) and NOT refactoring everything from ground up - because it is not directly related to the roi_align_rotated op itself. As such, I believe my proposition is better to yours and will keep it - honestly I don't see any drawbacks of having a const getter that defines a number directly from op specification and none of you were able so far to give me convincing explanation.

Reply:
There no need to add public member for such values or refactor the code.
The separation of shape inference has been done some time ago intentionally and consequence are known and accepted.

If there is possible to improve and make it more integrated, prepare ticket/PR with concept and if that give improvement
then it can be propagated to other operators.

Reply:
The need to keep such public members is a consequence of of the mixing validation and shape(along with separation shape inference) in one place. If you say that was done intentionally and consequences are known and accepted - then please do not block this PR, since once again - t**his is the one of the consequences of such design.**